### ISSUE TITLE
Giving Map<String, dynamic> with non-String || Iterable<String> to Uri.https() as queryParameters throws Exception

### ISSUE THREAD

**Issue by gysipos** [dart-lang/sdk#46251]
This tracker is for issues related to:
* Dart core libraries: uri.dart

* Dart SDK Version (`dart --version`): 2.13.1
 using MacOSX
 using Flutter

If you try to create a Uri.https() with Map<String, dynamic> queryParameters with values other than String or Iterable<String> it will fail to parse.

I think the source of the issue will be around here:

uri.dart:2163
`    queryParameters.forEach((key, value) {
      if (value == null || value is String) {
        writeParameter(key, value);
      } else {
        Iterable values = value;
        for (String value in values) {
          writeParameter(key, value);
        }
      }
    });`

It looks like they only manage cases where the value is either one of the above mentioned types. I got a hunch that someone even thought about this, since the declaration of the function looks like this:

uri.dart:2139
`  static String? _makeQuery(String? query, int start, int end,
      Map<String, dynamic /*String|Iterable<String>*/ >? queryParameters)`

They have the correct types in a comment right next to the dynamic!

**Comment by lrhn** [comment]
That is correct. It's not documented, and it should be.
Likely the documentation wasn't updated when we started allowing multiple strings as a valid value, before that the type of `Map<String, String>` was sufficient

**Comment by jamesderlin** [comment]
The documentation for [the default `Uri` constructor](https://api.dart.dev/stable/dart-core/Uri/Uri.html) explains what's expected for `queryParameters`, but the `Uri.http`/`Uri.https` documentation don't mention any of that (nor do they refer readers to the constructor for the default constructor).

**Comment by jamesderlin** [comment]
Additionally, I think `Uri` should generate a better error message instead of assuming that `queryParameters` is an `Iterable` if it isn't a `String`.  The current failure mode is very confusing.

**Comment by jamesderlin** [comment]
And, of course, another option would be to make `queryParameters` friendlier and less reliant on documentation by making it automatically call `.toString()` on non-`Iterable` `Map` values.

**Comment by lrhn** [comment]
Agree with everything until the last one. Calling `toString` on values can hide bugs. If you put in something which was intended to be an iterable, but you did an off-by-one error and ended up with something else, you'll immediately be told if we throw. If we just do `toString`, it might "work" for a while.

For values that are dynamically typed, so you don't get static type checking help to find bugs, I find that being very strict about which values you accept, and validating early and often, is usually better in the long run.

**Comment by gysipos** [comment]
I agree. The best improvement for me would be a more informative Exception, telling me about the only expected values being String or Iterable<String>. That way i'd have a clear and direct explanation in front of me about the issue that just happened. The next best thing would be for me to have a class for the arguments, and queryParams should take a Map<String, QueryParamValue>? like so:

`
class QueryParamValue {
  final String? one; //name up for debate
  final Iterable<String>? multiple; //name up for debate

  QueryParamValue({this.one, this.multiple})
      : assert(one == null && multiple == null,
            "one or multiple should be a non-null value"),
        assert(one != null && multiple != null,
            "only one of [one] or [multiple] should be used, not both");
}
`
The third best helpful solution from me would be the better documentation on dart.dev as @jamesderlin  mentioned.

I agree with @lrhn on the `.toString()` being problematic. Either make the unexpected not doable at all, or give me a clear indication about what the unexpected thing was, but don't try to figure out what i could have meant, that would open a whole new box of possible exceptions, making the whole issue more complicated and potentially confusing.

**Comment by Craftplacer** [comment]
Any update on this? I encountered this twice today before figuring it out.
